spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/NettyRoutingFilter.java

```diff
@@ -151,10 +152,9 @@ public int getOrder() {
 				});
 
 		if (properties.getResponseTimeout() != null) {
-			//TODO: figure out how to make this a 504
 			responseFlux = responseFlux.timeout(properties.getResponseTimeout(),
 					Mono.error(new TimeoutException("Response took longer than timeout: " +
-							properties.getResponseTimeout())));
+							properties.getResponseTimeout()))).onErrorMap(TimeoutException.class, th -> new ResponseStatusException(HttpStatus.GATEWAY_TIMEOUT, null, th));
```

Comment:
Technically, here one may say "Hey, why don't we get rid of `TimeoutException` altogether then"?

The answer is: it is a *very* convenient "marker" exception, which I already make use of in #555. I like the fact that, by not reusing `java.util.concurrent.TimeoutException`, one can tell apart this type of timeout from one issued, for example, by a filter using `java.util.concurrent.Future` for whatever dastardly reasons.